From 380d58ced938a57baa2e0696403d5826fd0829d5 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Stefan=20D=C3=B6singer?= <stefan@codeweavers.com>
Date: Wed, 10 Apr 2013 19:10:41 +0200
Subject: [PATCH 063/127] wined3d: Prevent the command stream from running
 ahead too far
Reply-To: wine-devel <wine-devel@winehq.org>

---
 dlls/wined3d/cs.c              | 8 ++++++++
 dlls/wined3d/wined3d_private.h | 9 +++------
 2 files changed, 11 insertions(+), 6 deletions(-)

diff --git a/dlls/wined3d/cs.c b/dlls/wined3d/cs.c
index 41b1cc4..9e47019 100644
--- a/dlls/wined3d/cs.c
+++ b/dlls/wined3d/cs.c
@@ -423,6 +423,8 @@ static UINT wined3d_cs_exec_present(struct wined3d_cs *cs, const void *data)
             src_rect, dst_rect, dirty_region, op->flags,
             cs->state.fb.depth_stencil);
 
+    InterlockedDecrement(&cs->pending_presents);
+
     return sizeof(*op);
 }
 
@@ -431,6 +433,7 @@ void wined3d_cs_emit_present(struct wined3d_cs *cs, struct wined3d_swapchain *sw
         const RGNDATA *dirty_region, DWORD flags)
 {
     struct wined3d_cs_present *op;
+    LONG pending;
 
     op = cs->ops->require_space(cs, sizeof(*op));
     op->opcode = WINED3D_CS_OP_PRESENT;
@@ -454,7 +457,12 @@ void wined3d_cs_emit_present(struct wined3d_cs *cs, struct wined3d_swapchain *sw
     }
     op->flags = flags;
 
+    pending = InterlockedIncrement(&cs->pending_presents);
+
     cs->ops->submit(cs);
+
+    while (pending > 1)
+        pending = InterlockedCompareExchange(&cs->pending_presents, 0, 0);
 }
 
 static UINT wined3d_cs_exec_clear(struct wined3d_cs *cs, const void *data)
diff --git a/dlls/wined3d/wined3d_private.h b/dlls/wined3d/wined3d_private.h
index 116e80e..3bc68b7 100644
--- a/dlls/wined3d/wined3d_private.h
+++ b/dlls/wined3d/wined3d_private.h
@@ -2503,12 +2503,9 @@ struct wined3d_cs
     DWORD tls_idx;
     struct wined3d_cs_list free_list;
     struct wined3d_cs_list exec_list;
-    /* FIXME: We should throttle presents, to prevent the application from
-     * getting too far ahead of the GPU. 2 or 3 frames is probably still
-     * acceptable. */
-#if 0
-    UINT present_idx;
-#endif
+
+    LONG pending_presents;
+
     struct wined3d_device *device;
     struct wined3d_surface *onscreen_depth_stencil;
     struct wined3d_state state;
-- 
1.8.1.5

